demos/adjoint/adjoint_demo.py

```diff
@@ -0,0 +1,342 @@
+# # Add the dataset to the plotter
+# plotter.add_mesh(dataset, scalars='Observed Temperature', cmap='coolwarm')
+# # Adjust the camera position
+# plotter.camera_position = [(0.5, 0.5, 2.5), (0.5, 0.5, 0), (0, 1, 0)]
+# # Show the plot
+# plotter.show()
+# -
+
+# The inverse code
+# ---------------------------------------
+#
+# The novelty of using the overloading approach provided by pyadjoint is that
+# we need minimal changes to our script in order to enable the inverse capabalities of G-ADOPT.
+# To turn on adjoint, one simply starts by importing
+# the inverse module to enable all taping functionality from pyadjoint.
+
+
+# Doing so will turn all Firedrake's objects to overloaded types, in a way
+# that any UFL operation will be annotated and added to the tape, unless
+# specified otherwise. To make sure that the tape is cleared of any
+# previous operations:
+
+# +
+# get the working tape
+tape = get_working_tape()
+# clear the tape from any previous annotation
+tape.clear_tape()
+# -
+# + tags=[active-ipynb]
+# # print all the blocks
+# print(tape.get_blocks())
+# -
+
+# From here onwards, every operation will be done with minimal differences with
+# respect to our forward code. Under the hood, however, the tape will be populated
+# by *blocks* that know the dependencies.
+# Knowing the mesh was loaded above, we continue very similar to our base case.
+
+# +
+# Set up function spaces for the Q2Q1 pair.
+V = VectorFunctionSpace(mesh, "CG", 2)  # Velocity function space (vector)
+W = FunctionSpace(mesh, "CG", 1)  # Pressure function space (scalar)
+Q = FunctionSpace(mesh, "CG", 2)  # Temperature function space (scalar)
+Z = MixedFunctionSpace([V, W])  # Mixed function space
+
+# Test functions and functions to hold solutions:
+z = Function(Z)  # A field over the mixed function space Z
+u, p = split(z)  # Returns symbolic UFL expression for u and p
+z.subfunctions[0].rename("Velocity")
+z.subfunctions[1].rename("Pressure")
+
+Ra = Constant(1e6)  # Rayleigh number
+approximation = BoussinesqApproximation(Ra)
+
+# Define time-stepping parameters.
+delta_t = Constant(4e-6)  # Constant time step
+timesteps = int(temperature_timestepping_info["index"][-1])  # number of timesteps from forward
+
+T = Function(Q, name="Temperature")
+
+# Nullspaces for the problem (constant pressure).
+Z_nullspace = create_stokes_nullspace(Z, closed=True, rotational=False)
+
+# Stokes and Energy boundary conditions - All free-slip for this problem
+stokes_bcs = {
+    bottom_id: {"uy": 0},
+    top_id: {"uy": 0},
+    left_id: {"ux": 0},
+    right_id: {"ux": 0},
+}
+temp_bcs = {
+    bottom_id: {"T": 1.0},
+    top_id: {"T": 0.0},
+}
+# Energy and Stokes solver
+energy_solver = EnergySolver(T, u, approximation, delta_t, ImplicitMidpoint, bcs=temp_bcs)
+stokes_solver = StokesSolver(z, T, approximation, bcs=stokes_bcs, nullspace=Z_nullspace, transpose_nullspace=Z_nullspace, constant_jacobian=True)
+# -
+
+# Define the Control Space
+# ========================
```

Comment:
I think the header levels need to be properly hierarchical, or it will break the table of contents in the notebook rendering. That is, this should be level 2, because the document title is level 1.

Reply:
Done! Thanks!